### PR TITLE
Actively abort abandoned Txns in (*TxnCoordSender).heartbeat, #3035

### PR DESCRIPTION
**Pull request by bg451**
Whenever a heartbeat considers a `Txn` abandoned, it sends out a `roachpb.EndTransactionRequest` to actively abort the transaction and its intents per #3035

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3108)

### CODE REVIEW

File: kv/txn_coord_sender.go [hunk 1]
```diff
@@ -600,7 +600,25 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
...
 	if !proceed {
+		// Actively abort the transaction and its intents since we assume it's abandoned
+		ab := &roachpb.EndTransactionRequest{
+			Commit:  false,
+			Intents: txnMeta.intents(),
+		}
+		ba.Add(ab)
+		if _, err := tc.wrapped.Send(ctx, ba); err != nil {
+			if log.V(1) {
+				log.Warningf("heartbeat abort to %s failed: %s ", txn, err)
```
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: use `abort due to inactivity failed for %s: %s`.

File: kv/txn_coord_sender.go [hunk 2]
```diff
@@ -600,7 +600,25 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
...
 	tc.Unlock()
+
+	ba := roachpb.BatchRequest{}
+	ba.Timestamp = tc.clock.Now()
+	ba.CmdID = ba.GetOrCreateCmdID(ba.Timestamp.WallTime)
+	ba.Txn = txn.Clone()
+
 	if !proceed {
+		// Actively abort the transaction and its intents since we assume it's abandoned
+		ab := &roachpb.EndTransactionRequest{
```
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: s/ab/et/

File: kv/txn_coord_sender.go [hunk 3]
```diff
@@ -600,7 +600,25 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
...
+			Commit:  false,
+			Intents: txnMeta.intents(),
```
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: you'll need to lock around here and atomically remove from `tc.txns`. Something like

```go
var intents []roachpb.Intent
tc.Lock()
if txnMeta, ok := tc.txns[id]; ok {
  delete(tc.txns, id)
  intents := txnMeta.intents()
}
tc.Unlock()

File: kv/txn_coord_sender.go [hunk 4]
```diff
@@ -600,7 +600,25 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
 	// so grab a copy before unlocking.
 	txn := txnMeta.txn
 	tc.Unlock()
+
+	ba := roachpb.BatchRequest{}
+	ba.Timestamp = tc.clock.Now()
+	ba.CmdID = ba.GetOrCreateCmdID(ba.Timestamp.WallTime)
```
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: Try rebasing, I think `CmdID` has since been removed.

Comment (bg451) [inline kv/txn_coord_sender.go]: WIll do

File: kv/txn_coord_sender.go (hunk 3 above)
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: sorry, scratch the `delete` above (that's already done in the `defer unregisterTxnLocked(id)` above, and I think we don't have to do it atomically.

**Comment by tschottdorf** [comment]
take a look at `TestTxnCoordSenderHeartbeat` and `TestTxnCoordSenderGC` for testing inspiration. You can probably augment the latter.

File: kv/txn_coord_sender.go (hunk 3 above)
Comment (tschottdorf) [inline kv/txn_coord_sender.go]: You should also consider not using the wrapped sender for `EndTransaction`, but instead sending it through `TxnCoordSender` itself. That would save you a lot of special casing.

Something like

`tc.stopper.RunAsyncTask(func() { tc.Send(....) })`.

**Comment by bg451** [comment]
@tschottdorf How's this?

File: kv/txn_coord_sender.go [hunk 5]
```diff
@@ -600,15 +600,30 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
...
+		// Actively abort the transaction and its intents since we assume it's abandoned
+		et := &roachpb.EndTransactionRequest{
+			Commit:  false,
+			Intents: txnMeta.intents(),
+		}
+		ba.Add(et)
+		tc.stopper.RunAsyncTask(func() {
+			if _, err := tc.Send(ctx, ba); err != nil {
+				if log.V(1) {
+					log.Warningf("abort due to inactivty for %s failed: %s ", txn, err)
```
Comment (tamird) [inline kv/txn_coord_sender.go]: s/inactivty/inactivity/

**Comment by bg451** [comment]
Huh, not quite sure what to think of the test failure. It's a nil pointer dereference in `TestReplicateRogueRemovedNode` and doesn't seem to have anything to do with the pr. Should I open a new issue for the test failure?

**Comment by bdarnell** [comment]
There's already an issue: #2880. I've restarted your build.

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[kv/txn_coord_sender.go, line 597 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3108#-K3VdpD1RVFI4h9W6Usd:-K3VdpD2eGFxBwfdE2Nn:961315352) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c2fcae13a1032ce225c5f9dced2e99d80dba5a/kv/txn_coord_sender.go#L597)):_
you need to grab the intents here and then call `txnMeta.keys.Clear()` to avoid data races.

---

_[kv/txn_coord_sender.go, line 609 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3108#-K3Vf1izZGvlTMB95Um1:-K3Vf1izZGvlTMB95Um2:1222012389) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c2fcae13a1032ce225c5f9dced2e99d80dba5a/kv/txn_coord_sender.go#L609)):_
add punctuation at the end.

---

_[kv/txn_coord_sender.go, line 616 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3108#-K3VeXeUIphCLXm2Y_jQ:-K3VeXeUIphCLXm2Y_jR:1689035979) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c2fcae13a1032ce225c5f9dced2e99d80dba5a/kv/txn_coord_sender.go#L616)):_
if you use `tc.Send`, you're not allowed to supply the intents manually. Probably best to just use the wrapped sender for now.

---

_[kv/txn_coord_sender_test.go, line 432 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3108#-K3VdPAhmSdycsTQWwME:-K3VdPAhmSdycsTQWwMF:382088299) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c2fcae13a1032ce225c5f9dced2e99d80dba5a/kv/txn_coord_sender_test.go#L432)):_
can you refactor this to use `util.SucceedsWithin`?

---

_[kv/txn_coord_sender_test.go, line 445 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3108#-K3Vcc2EITIyOCv-v6d2:-K3Vcc2EITIyOCv-v6d3:561513380) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c2fcae13a1032ce225c5f9dced2e99d80dba5a/kv/txn_coord_sender_test.go#L445)):_
you won't be able to send `EndTransaction` through `TxnCoordSender` after it's already removed the transaction from its state (so whatever error you're getting at the moment, it's not the right one). You're going to have to send straight through the wrapped sender (in this test, `s.Sender.wrapped` should do the trick).

Assert on the error. You should see a `*roachpb.TransactionAbortedError`:

```go
_, err := ....
if _, ok := err.(*roachpb.TransactionAbortedError); !ok {
  return err // assuming you refactored as suggested above
}
return nil
```

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3108)_

**Comment by kkaneda** [comment]
Made a small fix in #4614.

**Pull request closed without merging**